### PR TITLE
Add dict manipulation tools.

### PR DESCRIPTION
**Pull request by PaulSonOfLars**
This address issue #126, adding dictionary manipulation tools.

Specifically, this adds the following to buildozer:

`dict_add <attr> <(key:value)(s)>` Which does not overwrite existing values.
`dict_set <attr> <(key:value)(s)>` Which overwrites existing values.
`dict_remove <attr> <key(s)>` Which removes keys from the dictionary, removing the final dictionary if it is now empty.

### CODE REVIEW

**Comment by petemounce** [comment]
cc @devnev re: block-network?

File: edit/edit.go [hunk 1]
```diff
@@ -682,6 +682,18 @@ func DictionarySet(dict *build.DictExpr, key string, value build.Expr) build.Exp
 	return nil
 }
 
+// DictionaryGet looks for the key in the dictionary expression, and returns the
+// current value. If it is unset, it returns nil.
+func DictionaryGet(dict *build.DictExpr, key string) build.Expr {
+	for _, e := range dict.List {
+		kv, _ := e.(*build.KeyValueExpr)
```
Comment (pmbethe09) [inline edit/edit.go]: while it should be a KeyValueExpr, might as well inspect and continue if not

kv, ok := e.(*build.KeyValueExpr)
if !ok {
  continue
}

File: edit/buildozer.go [hunk 2]
```diff
@@ -454,6 +455,85 @@ func cmdCopyNoOverwrite(opts *Options, env CmdEnvironment) (*build.File, error)
...
+}
+
+// cmdDictRemove removes a key from a dict.
+func cmdDictRemove(opts *Options, env CmdEnvironment) (*build.File, error) {
+	attr := env.Args[0]
+	args := env.Args[1:]
+
+	thing := env.Rule.Attr(attr)
+	dictAttr, ok := thing.(*build.DictExpr)
+	if ok {
```
Comment (pmbethe09) [inline edit/buildozer.go]: if !ok, then this isn't a dict, so probably cleaner to invert-return
if !ok {
  return env.File, nil
}
for _, x := range args {

File: edit/buildozer.go [hunk 3]
```diff
@@ -454,6 +455,85 @@ func cmdCopyNoOverwrite(opts *Options, env CmdEnvironment) (*build.File, error)
...
+			// should errors here be flagged?
+			DictionaryDelete(dictAttr, x)
+			env.Rule.SetAttr(attr, dictAttr)
+		}
+	}
+
+	// If the removal results in the dict having no contents, delete the attribute (stay clean!)
+	if dictAttr == nil || len(dictAttr.List) == 0 {
+		if env.Rule.DelAttr(attr) != nil {
+			return env.File, nil
```
Comment (pmbethe09) [inline edit/buildozer.go]: this seems like clutter
if dictAttr == nil || len(dictAttr.List) == 0 {
  env.Rule.DelAttr(attr)
}
return env.File, nil

**Comment by PaulSonOfLars** [comment]
We need a way of calling function/methods from inside dicts; would it be acceptable to use a `dict_add_func` method for adding function calls to dictionaries? or is there some syntactic sugar already present to denote "this is a function" vs "this is a string"?

**Comment by petemounce** [comment]
Hi there - @PaulSonOfLars has moved on from improbable.io. I still work there. I notice the CLA check fails, but I'm pretty sure we're covered by a company-wide CLA. Is there anything I can do to unblock and get this merged?

**Comment by laurentlb** [comment]
@petemounce Can you create a new PR?

**Comment by petemounce** [comment]
@laurentlb - sure, done in #421.

**Comment by PaulSonOfLars** [comment]
Closing, since this was merged with #421. Big thanks to @petemounce taking over the CLA for me!

**Pull request closed without merging**